semantic/mesh/roof/mesh-repair.cpp

```diff
@@ -233,6 +238,153 @@ Mesh removeUnusedVertices(Mesh& mesh)
+                    isValid = false;
+                }
+
+                for (const auto& nI : edgeMap.at(e)) {
+                    if (nI == fI) { continue; }
+                    if (regions[nI] != -1) { continue; } // skip enqueued
+
+                    // Add neighbor to component & enqueue it
+                    regions[nI] = comp;
+                    q.push(nI);
+                }
+            }
+        }
+        if (isValid) { validRegions.push_back(comp); }
+        // no new faces to add
+        ++comp;
+    }
+}
+
+
+geometry::Mesh constructMeshPart(const geometry::Mesh& mesh, 
+                                 const std::vector<int>& regions, 
+                                 const int regionId)
+{
+
+    geometry::Mesh meshPart;
+    math::Points3::size_type vertexId = 0;
+    std::map<math::Point3, math::Points3::size_type> vertexIndexMap;
+
+    for (int id = 0; id < (int)regions.size(); ++id) {
+        if (regions[id] != regionId) { continue; }
+
+        geometry::Face face = mesh.faces[id];
+
+        math::Point3 a = mesh.a(face);
+        if (!vertexIndexMap.count(a)) {
+            meshPart.vertices.push_back(a);
+            vertexIndexMap[a] = vertexId;
+            vertexId += 1;
+        }
+
+        math::Point3 b = mesh.b(face);
+        if (!vertexIndexMap.count(b)) {
+            meshPart.vertices.push_back(b);
+            vertexIndexMap[b] = vertexId;
+            vertexId += 1;
+        }
+
+        math::Point3 c = mesh.c(face);
+        if (!vertexIndexMap.count(c)) {
+            meshPart.vertices.push_back(c);
+            vertexIndexMap[c] = vertexId;
+            vertexId += 1;
+        }
+
+        meshPart.addFace(vertexIndexMap[a],
+                         vertexIndexMap[b],
+                         vertexIndexMap[c],
+                         face.imageId);
+    }
+
+    return meshPart;
+}
+
+
+geometry::Mesh getValidComponents(const geometry::Mesh& mesh,
+                                  const std::vector<int>& regions,
+                                  const std::vector<int>& validRegions)
+{
+    geometry::Mesh validMesh;
+    for (int regionId : validRegions) {
+        geometry::Mesh meshPart(constructMeshPart(mesh, regions, regionId));
+        if (meshPart.volume() < 0.0001) {
+            LOG(info1) << "Mesh part " << regionId << " has zero volume. Removing.";
+            continue;
+        }
+        detail::append(validMesh, meshPart);
+    }
+    return validMesh;
+}
+
+
+geometry::Mesh removeNon2ManifoldParts(const geometry::Mesh& mesh)
+{
+    geometry::EdgeMap edgeMap = getNonManifoldEdgeMap(mesh);
+
+    // collect faces incident with non-manifold edge
+    std::set<Face::index_type> nonManifoldfaces;
+    for (auto& edge : edgeMap) {
+        if (edge.second.size() > 2) {
+            for (auto& fi : edge.second) {
+                nonManifoldfaces.insert(fi);
+            }
+        }
+    }
+
+    if ((int)nonManifoldfaces.size() == 0) { return mesh; }
```

Comment:
Pls, no C-cast. Either use _ctor_ cast like `int(...)` or full blown `static_cast` like `static_cast<int>(...)`. Not ti mention that there's no need to cast when comparing with zero.

However, something completely different belongs here:
```
if (nonManifoldfaces.empty()) { return mesh; }
```

Reply:
You are right,  changed.